rfcs/0014-dedicated-serving-cluster.md

```diff
@@ -0,0 +1,76 @@
+---
+feature: dedicated-serving-cluster
+authors:
+  - "Eric Fu"
+start_date: "2022/11/07"
+---
+
+# Dedicated Batch Serving Cluster
+
+## Summary
+
+Run a dedicated cluster to serve batch queries.
+
+## Motivation
+
+- **Different traffic pattern**. Batch queries, especially the high-concurrent point queries (aka. “online serving”), requires millisecond-level low latency and high availability, while streaming jobs focus more on high throughput and cost efficiency.
+- **Different access pattern on storage** of batch & streaming. Batch queries are issued by services on user side, while streaming operators always handle the changes from upstreaming system. It's possible that they have quite different distribution of access aka. hotspot. 
+- In some extreme use cases, the QPS on serving side could be 100K or even higher, which may result in severe **resource contention** (CPU, IOPS, etc.) against streaming jobs.
+
+To address these problems, here we propose to run a dedicated cluster to serve batch queries (hereafter referred to as “serving cluster”). Meanwhile, data source, sink and maintenance of MViews are still done in the primary streaming cluster.
+
+Additionally, this should be an **optional** feature. Some simple use cases, like dashboard, may not need to support such high concurrency. Besides, free tier RisingWave Cloud users are definitely not its target users.  
+
+## Design
+
+Decoupling barrier and checkpoint (reference: [Support barrier-checkpoint decoupling](https://singularity-data.quip.com/QLOaAegyeApR/Support-barrier-checkpoint-decoupling)) allows us to decouple the streaming barrier from the expensive checkpointing process, but also brings more complexity to us.  
+
+### Read Checkpointed Snapshots
+
+In the first stage, we can simply boot up several compute nodes without scheduling any streaming tasks on them. The meta service pushes the new storage versions to the compute nodes as usual, therefore, whenever a new storage version generated, the compute node will be informed.
+
+In the catalog, beside the existing `vnode -> parallel unit` mapping, we have to maintain another mapping of serving cluster for each table & materialized view. Below is an example, where “PU” refers to Parallel Unit. Apparently, this mapping should be updated once nodes joined or leaved the serving cluster. 
```

Comment:
How do we determine and reschedule this partition mapping? Should we have some heuristic strategies to replicate the hot spot more, or we simply scale with the cluster?

Reply:
That belongs to optimizations. For now, I prefer to assume the traffic can always be well distributed to all vnodes (so no "hot" vnode)